---
title: "clustermap"
author: "Ole Christian Lingjærde & Chloé B. Steen"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{clustermap}
  %\VignetteEngine{knitr::knitr}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
Clustermap is a collection of R functions for hierarchical clustering and heatmap visualisation of numerical data matrices. It is designed with two specific aims in mind: ease of use and flexibility. This manual will present some general features of `clustermap`, provide a more in-depth descriptions of the functions implemented in `clustermap` and finally, guide the user through a number of examples of increasing complexity.

## 1. General notes

###Data format
`clustermap` takes in any numeric matrix as input dataset. 

###The `clustermap` approach
Some of the functions in this package require other functions to have been performed in advance. For example, to display dendrograms along the sides of the heatmap, you must first have performed clustering.

Stick to this order and you will be safe:

```
plot.init -> 
    hcluster ->   
       subclust -> 
            plot.hmap, set.color, plot.cbar, plot.tree -> 
                                                   plot.ckey
```
`plot.init` erases any memory of past calls made to any of the functions in the package, so you get a fresh start. `hcluster` is used to cluster either columns or rows and you specify what type of clustering you would like (linkage, distance, etc). `subclust` may not be required in your analysis; what it does is to determine how many distinct clusters are actually present in the preceeding clustering. And then comes a series of plot functions to show the heatmap, any color bars you may want to show on the left/right or bottom/top of the heatmap, the cluster trees, and finally a color key (`plot.ckey`) that explains what values the heatmap colors represent.

###Plot to file
Plotting to a file is generally much faster than plotting to a graphics window. It is recommended that you start the session by opening a file as a plot device using e.g. pdf("filename.pdf") and then at the end of the session use dev.off() to close the device. This will be shown in the examples given further below.

###Dependencies
`clustermap` loads the packages `gclus` and `clusterGenomics`. These packages may be installed separately using the following commands:
```
install.packages("gclus")
install.packages("clusterGenomics")
```
or using the Package Installer inside R, found under the menu item 'Packages & Data'. 

###Colors
To change the colors used in the heatmap, provide your preferred color scale as an argument to plot.hmap, e.g.

```
plot.hmap(X, colorscale="green-white-red")
```
to use a color scale ranging from green (for negative values) via white (for value 0) to red (for positive values). 

You may provide any three color names supported by R; just a few examples:

* `blue-white-red`
* `green-black-red`
* `lightgrey-grey-darkgrey`
* `magenta-cyan-pink`

To obtain a textual overview of all colors available, give the command colors() or to see a visual overview over colors available, see <http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf>.

## 2. Functions in `clustermap`

###Summary of `clustermap` functions
`plot.init`		Call this first to set up correct margins etc

`plot.hmap` 	Plot heatmap using your preferred colors

`plot.cbar` 	Add color bar to one of the sides (to add extra info)

`plot.tree`		Add column or row dendrogram to a heatmap

`hcluster`     	Perform hierarchical clustering of rows or columns

`subclust`		Guided or automatic identification of subclusters

`get.subclust`  Get subtype labels (for original row/col ordering of X)

`set.color`     Set color coding for accompanying data

`plot.hmap.key` Plot color key for heatmap

`plot.cbar.key`	Plot color key for color bar (annotation bars)

`plot.text`		Add text to one of the sides (for row/column labels)

`plot.silhouette`  Silhouette plot

`blocks`		Generate an example data set with 110 rows and 40 columns

###plot.init()
`plot.init` is used to initiate a plot. 

The purpose is to make room on the four sides of the heatmap for additional elements such as cluster dendrograms, color bars with additional info, and labels. Supply as arguments the sides where you want such elements, with the coding 

side=1 (below)
side=2 (left)
side=3 (top)
side=4 (right)

If you want a tree (=dendrogram) on the left and top, let tree=c(2,3).
If you want text labels on the right, let text=4. 
If you want color bars below the plot, let cbar=1. 

Examples:
```
`plot.init()`                # Dendrogram on the left and on the top
`plot.init(tree=c())`        # Make no room for additional elements
`plot.init(tree=3, text=4)`  # Dendrogram on the top, labels on the right
```
The last two arguments of `plot.init` (`inner` and `outer`) do not overrule the other parameters, but allows the user to add (or even subtract) a specific amount of space in the inner and outer margins on a specified side of the heatmap. They may be used alone, or in combination with other parameters. They should always be vectors of length 4 giving the desired margin extension for each of the four sides. An example of use would be to extend an inner margin by a certain amount because we plan to show many bar plots of accompanying data there that we want to give some space. Note that inner=c(0.4,0.4,0.4,0.4) means that each inner margin should be extended by a length corresponding to 0.4 times the total length (in that direction) of the heatmap itself.

###subclust()
Use the function `subclust` to identify subclusters in a dendrogram. The effect is that later calls to `plot.tree` will show the subclusters in different colors. You can either provide the desired number of clusters using the argument `k`, or you can set `k=NA` to automatically estimate the number of cluster with one of two algorithms: gap (Tibshirani et al (2001), J Roy Statist Soc B, 63: 411-423) or PART (Nilsen et al (2012), Stat Appl Genet Mol Biol, 12: 637-652).

Using gap (Tibshirani et al., 2001), the result is an estimate of the optimal 'flat cut' of the dendrogram, while using PART (Nilsen et al., 2012), the result is an estimate of the optimal 'nonflat cut' of the dendrogram. In essence the PART algorithm is an extension of gap that applies gap recursively on subclusters previously identified. `B` and `min.size` are used by PART (only if k=NA) and determines the number of permutations (`B=25` is low, but computation time increases with increasing `B`; consider `B=200` or higher for high-quality estimates) and the least number of elements in a single cluster (`min.size=5` is quite arbitrary and you may want to change this). Note that PART may call some individuals as outliers and these will be shown in a separate color but may not be close to each other and will stick out as "clusters" of size potentially smaller than `min.size`.

###get.subclust()
`subclust` retrieves a vector of cluster labels after identification of subclusters. By default, labels are returned in the same order as shown in the cluster tree (`order="tree"`). To obtain labels in the original ordering in the input data X, use `order="orig"`. 

By default, the function returns a vector of cluster labels. If you supply row/column labels using the argument `labels` the function returns a data frame with two columns, the first being the cluster labels and the second being the supplied row/column labels.

###plot.hmap()
`plot.hmap()` plota heatmaps of a numerical data matrix X. It is recommended that you print to a file (e.g. a pdf-file) instead to increase speed (and use fast=F). Use the argument colorscale to control how numerical values are represented as colors in the heatmap. The argument should be three (valid R) colors separated by hyphens.
Some examples: "green-black-red", "yellow-black-blue", "black-white-black" (to ignore signs), and "white-grey-darkgrey".

###plot.text()
`plot.text()` is used to add labels to the sides of the heatmap. The first argument should be a vector of the same length as the number of columns in the heatmap (when side=1 or side=3), and a vector of the same length as the number of rows in the heatmap (when side=2 or side=4). The size of the text can be changed with the `cex` argument. Text labels can be truncated to a certain maximal length (for example 4 characters) with the `maxchar` argument. 

## 3. Examples using `clustermap`

###Set up everything
The following code sets up a simple example data set that can be used to run the examples below. It creates a hypothetical gene expression matrix consisting of 110 genes as rows, and 40 patient samples as columns. The `y1` vector contains information about cancer subtype of each sample. The `y2` vector contains information about stage, and finally the `y3` vector about immune score. 

```
library(clustermap) # Make package functions available
tmp = blocks() # Contains the 110x40-matrix X and 40-vectors y1,y2,y3
X = tmp$X
y1 = tmp$y1 
y2 = tmp$y2
y3 = tmp$y3
```

###How to create a heatmap
The following code creates a simple heatmap with no clustering, using the default color-scale ("blue-white-red") and adds the key for the color scale. 
```
pdf("Example1.pdf")
plot.init()
plot.hmap(X)
plot.hmap.key()
dev.off()
```

###Heatmap and clustering of columns
The following example shows how to cluster columns in a heatmap using the `hcluster()` function, and how to add a tree (dendogram) resulting from the column-wise clustering on the top side of the heatmap (`side=3`).
```
pdf("Example2.pdf")
plot.init(tree=3)
hcluster(X, clust="col", distance="pearson", linkage="complete")
plot.hmap(X) 
plot.tree(side=3)
plot.hmap.key()
frame()
dev.off()
```

###Heatmap and clustering of rows
This example is the same as the previous one, except that this time the clustering is performed row-wise, and the tree is added on the left side of the heatmap (`side=2`). 
```
pdf("Example3.pdf")
plot.init(tree=2)
hcluster(X, clust="row", distance="euclidean", linkage="complete")
plot.hmap(X)
plot.tree(side=2)
plot.hmap.key()
dev.off()
```
###Heatmap and clustering of rows and columns
Similar example as above, where the clustering is performed both row-wise and column-wise, and trees are added on the left side and on the top of the heatmap. 

```
pdf("Example4.pdf")
plot.init(tree=c(2,3))
hcluster(X, clust="row", distance="euclidean", linkage="complete")
hcluster(X, clust="col", distance="euclidean", linkage="complete")
plot.hmap(X)
plot.tree(side=2)
plot.tree(side=3)
plot.hmap.key()
dev.off()
```

###Heatmap, clustering of rows and columns, and colorbar below heatmap
The following example adds a color bar at the bottom of heatmap, with a discrete color scales made up of 5 colors, using the `y1` vector. The color bar is labeled "Subtype". 

```
pdf("Example5.pdf")
plot.init(tree=c(2,3), cbar=1, text=4)
hcluster(X, clust="row", distance="euclidean", linkage="complete")
hcluster(X, clust="col", distance="euclidean", linkage="complete")
plot.hmap(X)
plot.tree(side=2)
plot.tree(side=3)
z1 = set.color(y1, label="Subtype", type="discrete", color=c("red", "pink", "darkblue", "lightblue", "green"))
plot.cbar(z1, side=1)
plot.hmap.key()
plot.cbar.key()
dev.off()
```

###How to add multiple color bars below a heatmap
The following example shows how to add multiple color bars below a heatmap, to provide information about cancer subtype (`z1`), stage (`z2`) and immune score (`z3`). This is done by providing multiple `set.color` objects, here called `z1`, `z2` and `z3` to the `plot.cbar` function. 

```
pdf("Example6.pdf")
plot.init(tree=c(2,3), cbar=1, text=4)
hcluster(X, clust="row", distance="euclidean", linkage="complete")
hcluster(X, clust="col", distance="euclidean", linkage="complete")
plot.hmap(X)
plot.tree(side=2)
plot.tree(side=3)
z1 = set.color(y1, label="Subtype", type="discrete", color=c("red", "pink", "darkblue", "lightblue", "green"))
z2 = set.color(y2, label="Stage", type="discrete", color=rainbow(8))
z3 = set.color(y3, label="Immune score", type="continuous", color="white-blue")
plot.cbar(z1, z2, z3, border="black", side=1)
plot.hmap.key()
dev.off()
```

###Add row labels to a heatmap
The following code adds the rownames of the dataset `X` as row labels to the heatmap, using the function `plot.text`. The size the text is set using the argument `cex`, and the placement of the text using the argument `side`, where here `4` refers to the right side of the heatmap. 

```
pdf("Example7.pdf")
plot.init(tree=c(2,3), cbar=3, text=4)
hcluster(X, clust="row", distance="euclidean", linkage="complete")
hcluster(X, clust="col", distance="euclidean", linkage="complete")
plot.hmap(X)
plot.tree(side=2)
plot.tree(side=3)
z1 = set.color(y1, label="Subtype", type="discrete", color=c("red", "pink", "darkblue", "lightblue", "green"))
z2 = set.color(y2, label="Stage", type="discrete", color=palette())
z3 = set.color(y3, label="Immune score", type="continuous", color="black-yellow")
plot.cbar(z1, z2, z3, border="black", side=3)
plot.text(rownames(X), cex=0.3, side=4)
plot.hmap.key()
dev.off()
```

###Add columns labels to a heatmap
Similarly to the example above, labels can be added to columns using the `plot.text()` function. The size the text is set using the argument `cex`, and the placement of the text using the argument `side`, where here `3` refers to the top of the heatmap. 

```
pdf("Example8.pdf")
plot.init(tree=c(2,3), cbar=3, text=c(1,4))
hcluster(X, clust="row", distance="euclidean", linkage="complete")
hcluster(X, clust="col", distance="euclidean", linkage="complete")
plot.hmap(X)
plot.tree(side=2)
plot.tree(side=3)
z1 = set.color(y1, label="Subtype", type="discrete", color=c("red", "pink", "darkblue", "lightblue", "green"))
z2 = set.color(y2, label="Stage", type="discrete", color=palette())
z3 = set.color(y3, label="Immune score", type="continuous", color="black-yellow")
plot.cbar(z1, z2, z3, border="black", side=3)
plot.text(colnames(X), cex=0.3, side=1)
plot.hmap.key()
dev.off()
```

## Heatmap, clustering of rows and columns, define subclusters
The following example illustrates how to cut a dendrogram into `k=3` clusters and show in plot. The function `subclust()` is used to identify 3 clusters column-wise (`clust="col"`). To highlight the new tree with the colored clusters, the `lwd` argument in `plot.tree` is set to `2` to make the branches thicker. 

```
pdf("Example9.pdf")
plot.init(tree=c(2,3))
hcluster(X, clust="row", distance="euclidean", linkage="complete")
hcluster(X, clust="col", distance="euclidean", linkage="complete")
subclust(3, clust="col")
plot.hmap(X)
plot.tree(side=2)
plot.tree(side=3, lwd=2)
plot.hmap.key()
dev.off()
```

##Heatmap, clustering of rows and columns, estimate subclusters
The example above used a set number of clusters (`k=3`). But the number of clusters can also be automatically estimated using for example the GAP algorithm, which is default in the `subclust()` function. The other option is to use the PART algorithm, not shown here. The estimated clusters can then be visualized. Both row-wise and column-wise subclustering is shown. 

```
pdf("Example10.pdf")
plot.init(tree=c(2,3))
hcluster(X, clust="row", distance="euclidean", linkage="complete")
hcluster(X, clust="col", distance="euclidean", linkage="complete")
subclust(NA, clust="col")
subclust(NA, clust="row") 
plot.hmap(X)
plot.tree(side=2)
plot.tree(side=3)
plot.hmap.key()
dev.off()
```

###Heatmap, clustering of rows and columns, estimate subclusters
The previous example estimated the number of subclusters with the default algorithm gap. Here an example is shown where the number of subclusters is determined with the PART algorithm, set by the `method` argument of the subclust `function`. Both row-wise and column-wise subclustering is shown. 

```
pdf("Example11.pdf")
plot.init(tree=c(2,3))
hcluster(X, clust="row", distance="euclidean", linkage="complete")
hcluster(X, clust="col", distance="euclidean", linkage="complete")
subclust(NA, clust="col", method="part")
subclust(NA, clust="row", method="part")
plot.hmap(X)
plot.tree(side=2)
plot.tree(side=3)
plot.hmap.key()
dev.off()
```

###Heatmap, clustering of rows and columns, estimate subclusters, perform tests
After estimating subclusters, the user can test the association between additional info parameters and subclusters. This is done using the `plot.cbar()` function, which adds additional information to heatmap, and by specifying the `pvalue.method`, here `"fisher"`, a p-value showing the association between the subclusters and the color bar added is displayed on the heatmap. 

```
pdf("Example12.pdf")
plot.init(tree=c(2,3), cbar=3, text=4)
hcluster(X, clust="col", distance="euclidean", linkage="complete")
hcluster(X, clust="row", distance="euclidean", linkage="complete")
subclust(NA, clust="col", method="part")
subclust(NA, clust="row", method="part")
plot.hmap(X)
z1 = set.color(y1, label="Subtype", type="discrete", color=c("red", "pink", "darkblue", "lightblue", "green"))
z2 = set.color(y2, label="Stage", type="discrete", color=palette())
z3 = set.color(y3, label="Immune score", type="continuous", color="black-yellow")
plot.cbar(z1, z2, z3, pvalue=T, pvalue.method="fisher", border="black", side=3)
plot.tree(side=2)
plot.tree(side=3)
plot.hmap.key()
plot.cbar.key()
dev.off()
```

###Clustered heatmap where the clustering is derived from another data set 
Sometimes, it is useful to apply the clustering estimated using one dataset, to another dataset of the same size, for example is one wishes to follow patient samples over time, using the original subgrouping identified in the first samples. This is done by plotting a heatmap using another dataset with the `plot.hmap()` function, right after having performed the clustering on the original dataset with `hcluster()`. 

```
pdf("Example13.pdf")
plot.init(tree=c(2,3))
hcluster(X, clust="row", distance="euclidean", linkage="complete")
hcluster(X, clust="col", distance="euclidean", linkage="complete")
Xnew = tanh(X) # Xnew can be any numerical matrix of same size as X
plot.hmap(Xnew)
plot.tree(side=2)
plot.tree(side=3)
plot.hmap.key()
dev.off()
```

###Show three clustered heatmaps on the same page
Multiple heatmaps can be displayed on the same page using the `frame()` function. 

```
pdf("Example14.pdf")
xmax = max(2*X, na.rm=T)
par(mfrow=c(2,2))
panel.init(mfrow=c(2,2))
plot.init(tree=c(2,3), ckey=F, outer=c(0.1, 0.1, 0.1, 0.1))
hcluster(X, clust="row", distance="euclidean", linkage="complete")
hcluster(X, clust="col", distance="euclidean", linkage="complete")
plot.hmap(X, xmax=xmax)
plot.tree(side=2)
plot.tree(side=3)
plot.hmap.key()
frame()
plot.init(tree=c(2,3), ckey=F, outer=c(0.1, 0.1, 0.1, 0.1))
hcluster(X*2, clust="row", distance="pearson", linkage="complete")
hcluster(X*2, clust="col", distance="pearson", linkage="complete")
plot.hmap(X*2, xmax=xmax)
plot.tree(side=2)
plot.tree(side=3)
plot.hmap.key()
frame()
plot.init(tree=c(2,3), ckey=F, outer=c(0.1, 0.1, 0.1, 0.1))
hcluster(X/2, clust="row", distance="euclidean", linkage="average")
hcluster(X/2, clust="col", distance="euclidean", linkage="average")
plot.hmap(X/2, xmax=xmax)
plot.tree(side=2)
plot.tree(side=3)
plot.hmap.key()
frame()
dev.off()
```

###Cluster color image
Clustering is not restricted to gene expression data, here a more complex matrix from an image is clustered. 

```
library(magick)
pdf("Example15.pdf", width=7, height=2.5)
pict = bellagio()
panel.init(mfrow=c(1,3))
plot.init(ckey=F)
plot.hmap(pict$X)
frame()
frame()
plot.init(ckey=F)
plot.hmap(pict$Xrc)
frame()
frame()
plot.init(ckey=F)
Xcol = col2num(pict$Xrc, clust="col") # In each column replace color by (R,G,B)
Xrow = col2num(pict$Xrc, clust="row") # In each row replace color by (R,G,B)
hcluster(Xcol, clust="col", linkage="average") # Cluster columns using (R,G,B) values
hcluster(Xrow, clust="row", linkage="average") # Cluster rows using (R,G,B) values
plot.hmap(pict$Xrc, origin=1)
frame()
frame()
dev.off()
```

